Define GenericParamListRequest

CodaFi opened the pull request:
Formalize the lazy generic parameter list pattern into a request.

Also clean up a bunch of places where we were doing fun things to build ASTs around this pattern, squash the hacks that involved forcing these lists, and fix a bug in the nominal type decl request that could cause the generic parameters to be cyclically computed.

CodaFi committed:
Clean up the way Parse deals with GenericParamList

- The parser no longer mutates these in place
- Pass generic param lists off to constructors that will readily take them

CodaFi committed:
Define GenericParamListRequest

GenericParamListRequest formalizes the lazy generic parameter list construction pattern we were performing before.

CodaFi committed:
Fix the declcontext used to find the extended nominal type

This should not start searching at the extension.  Start looking one
level up.  Breaks a potential cycle where an extension's generic
parameters were subject to lookup while computing the nominal type.

CodaFi committed:
Drop a hack that forced generic params

CodaFi commented:
@swift-ci please test

slavapestov commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4828,7 +4827,8 @@ namespace {
         auto genericParams = *gpImportResult;
         if (genericParams) {
-          result->setGenericParams(genericParams);
+          result->getASTContext().evaluator.cacheOutput(
```
It seems that this comes up often enough that a setGenericParams() method is still useful. I would instead factor it so that GenericParamListRequest::cacheResult() calls setGenericParams().

slavapestov commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -3723,12 +3727,14 @@ class swift::DeclDeserializer {
-      extension->setGenericParams(genericParams);
+      // Set up the DeclContexts for the GenericTypeParamDecls in the list.
+      for (auto param : *genericParams)
+        param->setDeclContext(extension);
```
This is part of your GenericParamListRequest::cacheResult() so it's redundant here

CodaFi replied:
It's not redundant for the inner parameter lists - the old code was pumping `setGenericParams` just for this effect.

slavapestov replied:
Perhaps later we can investigate deserializing the generic parameter list after we deserialize its owner decl, giving generic parameter decls the right DeclContext upfront. At one point there were circularity issues though

slavapestov commented on lib/Serialization/Serialization.cpp:
```diff
@@ -3088,7 +3088,6 @@ class Serializer::DeclSerializer : public DeclVisitor<DeclSerializer> {
                                inheritedAndDependencyTypes);
 
-    const_cast<ProtocolDecl*>(proto)->createGenericParamsIfMissing();
     writeGenericParams(proto->getGenericParams());
```
Optional cleanup: we gain nothing by serializing or deserializing the generic parameter list of a protocol, since we can always re-derive it from scratch.

slavapestov reviewed (approved):
Nice cleanup!

CodaFi committed:
Update a flaky test

As we continue to requestify more parts of the compiler, the CHECK lines
here will continue to change.

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
OK, definitely going to land this one *first* before I have to rebase this *again*.

@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please python lint

CodaFi commented:
@swift-ci please smoke test macOS platform

CodaFi commented:
⛵️
